packages/flutter/test/widgets/nested_scroll_view_test.dart

```diff
@@ -676,8 +676,309 @@ void main() {
     await gesture.up();
     debugDefaultTargetPlatformOverride = null;
   });
+
+  testWidgets('NestedScrollView exposes scroll controllers',
+          (WidgetTester tester) async {
+    final GlobalKey<NestedScrollViewState> globalKey = GlobalKey();
+
+    await tester.pumpWidget(Directionality(
+      textDirection: TextDirection.ltr,
+      child: Localizations(
+        locale: const Locale('en', 'US'),
+        delegates: const <LocalizationsDelegate<dynamic>>[
+          DefaultMaterialLocalizations.delegate,
+          DefaultWidgetsLocalizations.delegate,
+        ],
+        child: MediaQuery(
+          data: const MediaQueryData(),
+          child: NestedScrollView(
+            key: globalKey,
+            body: Container(),
+            headerSliverBuilder: (_, __) => <Widget>[const SliverAppBar()],
+          ),
+        ),
+      ),
+    ));
+
+    expect(globalKey.currentState.innerController, isNotNull);
+    expect(globalKey.currentState.outerController, isNotNull);
+  });
+
+  // One scrollExtent should be smaller than the height of the toolbar
+  // to ensure that the inner scroll controller is not scrolled in that case.
+  for (final double scrollExtent in const <double>[
```

Comment:
As pointed out before: This test is very hard to understand because of the many conditions. It would be better to separate it out into individual test cases. Each test case should have a self-explaining name: e.g. "Scrolling by less then the header scroll extend does not scroll the inner body". That also removes the awkward comment on top here.